### PR TITLE
Add support for level 4 selectors in selector-max-specificity

### PR DESCRIPTION
**Pull request by Hypnosphi**
This also adds `ignoreSelectors` option.

> Which issue, if any, is this issue related to?

https://github.com/stylelint/stylelint/issues/2857#issuecomment-328304228

### CODE REVIEW

File: lib/rules/selector-max-specificity/README.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ Visit the [Specificity Calculator](https://specificity.keegan.st) for visual rep
 
 This rule ignores selectors with variable interpolation (`#{$var}`, `@{var}`, `$(var)`).
 
-This rule ignores selectors containing the `:not()` or `:matches()` pseudo-classes.
+This rule handles selectors containing the `:not()` or `:matches()` pseudo-classes per [spec](https://drafts.csswg.org/selectors/#specificity-rules).
```
Comment (jeddy3) [inline lib/rules/selector-max-specificity/README.md]: I think we can drop this line completely as the user can assume it supports level 4 selectors. These sentences are generally used to inform the user of _limitations_ of the rule, when dealing with specs.

File: lib/rules/selector-max-specificity/README.md [hunk 2]
```diff
@@ -79,3 +79,43 @@ div {}
...
+
+```css
+:local(.foo.bar)
+```
+
+```css
+:local(.foo, :global(.bar).baz)
+```
+
+The following patterns are still considered violations:
```
Comment (jeddy3) [inline lib/rules/selector-max-specificity/README.md]: You weren't to know this, but this must read exactly:

```
The following patterns are considered violations:
```

For the red/green background color highlighting to work on [the website](https://stylelint.io/user-guide/rules/color-no-invalid-hex/).

File: lib/rules/selector-max-specificity/index.js [hunk 3]
```diff
@@ -15,22 +18,97 @@ const messages = ruleMessages(ruleName, {
...
+          ignoreSelectors: [_.isString]
+        },
+        optional: true
+      }
+    );
     if (!validOptions) {
       return;
     }
 
+    // Calculate the specificity of a simple selector (type, attribute, class, id, or pseudos"s own value)
```
Comment (jeddy3) [inline lib/rules/selector-max-specificity/index.js]: Typo: pseudos"s -> pseudos's

File: lib/rules/selector-max-specificity/index.js [hunk 4]
```diff
@@ -15,22 +18,97 @@ const messages = ruleMessages(ruleName, {
...
+      const ownSpecificity =
+        ownValue === ":not" || ownValue === ":matches"
+          ? // :not and :matches don't add specificity themselves, but their children do
+            zeroSpecificity()
+          : simpleSpecificity(ownValue);
+
+      return specificitySum([ownSpecificity, maxChildSpecificity(node)]);
+    };
+
+    // Calculate the specificity of a node parsed bu `postcss-selector-parser`
```
Comment (jeddy3) [inline lib/rules/selector-max-specificity/index.js]: Typo:  parsed bu -> parsed by

**Comment by ntwb** [comment]
Changelog https://github.com/stylelint/stylelint/commit/fdc485e62c553c338b5e65742d3c176965e60932
> -   Added: Add support for level 4 evaluation context pseudo-classes in   `selector-max-specificity` ([#2857](https://github.com/stylelint/stylelint/issues/2857)).

**Comment by jeddy3** [comment]
FYI, I tweaked the changelog entry to:

> -   Added: `selector-max-specificity` support for level 4 evaluation context pseudo-classes

So that it [leads with the rule name](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/pull-requests.md).

The CHANGELOG now reads:

-   Added: `length-zero-no-unit` autofix ([#2861](https://github.com/stylelint/stylelint/issues/2861)).
-   Added: `selector-max-specificity` support for level 4 evaluation context pseudo-classes ([#2857](https://github.com/stylelint/stylelint/issues/2857)).
-   Fixed: `selector-pseudo-class-no-unknown` false positives when using chained pseudo-classes ([#2810](https://github.com/stylelint/stylelint/issues/2810)).
-   Fixed: `unit-no-unknown` false positives for spaceless multiplication and division in `calc()` functions ([#2848](https://github.com/stylelint/stylelint/issues/2848)).

**Pull request merged**